lib/pdk/module/templatedir.rb

```diff
@@ -37,17 +37,27 @@ class TemplateDir
       # @raise [ArgumentError] (see #validate_module_template!)
       #
       # @api public
-      def initialize(path_or_url, module_metadata = {}, init = false)
+      def initialize(uri, module_metadata = {}, init = false)
         unless block_given?
           raise ArgumentError, _('%{class_name} must be initialized with a block.') % { class_name: self.class.name }
         end
+        unless uri.is_a? PDK::Util::TemplateURI
+          raise ArgumentError, _('PDK::Module::TemplateDir.new must be initialized with a PDK::Util::TemplateURI, got a %{uri_type}') % { uri_type: uri.class }
+        end
 
-        if PDK::Util::Git.repo?(path_or_url)
-          @path = self.class.clone_template_repo(path_or_url)
-          @repo = path_or_url
+        if PDK::Util::Git.repo?(uri.git_remote)
+          # This is either a bare local repo or a remote. either way it needs cloning.
+          @path = self.class.clone_template_repo(uri)
+          temp_dir_clone = true
         else
-          @path = path_or_url
+          # if it is a local path & non-bare repo then we can use it directly.
+          # Still have to check the branch.
+          @path = uri.shell_path
+          # We don't do a checkout of local-path repos. There are lots of edge
+          # cases or user un-expectations.
+          PDK.logger.warn _("Repository '%{repo}' has a work-tree; skipping git reset.") % { repo: @path }
```

Comment:
The behaviour itself doesn't actually change, just the UX. I think if we just guard this `warn` so that it only happens if this is in fact a git repo and not just a local directory we shouldn't have any substantial behaviour change.